### PR TITLE
Remove assignments to ZST places instead of marking ZST return place as unused

### PR DESCRIPTION
**Pull request by erikdesjardins**
partially reverts #83118

requested by @tmiasko in https://github.com/rust-lang/rust/pull/83118#issuecomment-799692574

r? @oli-obk

### CODE REVIEW

File: compiler/rustc_mir/src/transform/instcombine.rs [hunk 1]
```diff
@@ -41,6 +44,28 @@ impl<'tcx, 'a> InstCombineContext<'tcx, 'a> {
         })
     }
 
+    /// Remove assignments to inhabited ZST places.
+    fn combine_zst(&self, source_info: &SourceInfo, kind: &mut StatementKind<'tcx>) {
+        match kind {
+            StatementKind::Assign(box (place, _)) => {
+                let place_ty = place.ty(self.local_decls, self.tcx).ty;
+                if let Ok(layout) = self.tcx.layout_of(self.param_env.and(place_ty)) {
+                    if layout.is_zst() && !layout.abi.is_uninhabited() {
```
Comment (erikdesjardins) [inline compiler/rustc_mir/src/transform/instcombine.rs:53]: Removing assignments of uninhabited ZSTs changed the error output for `ui/consts/const-eval/panic-never-type.rs` and `ui/consts/const-eval/panic-assoc-never-type/panic-assoc-never-type.rs`.

e.g.

```diff
---- [ui] ui/consts/const-eval/panic-assoc-never-type.rs stdout ----
diff of stderr:

15         = note: for more information, see issue #71800 <https://github.com/rust-lang/rust/issues/71800>
16         = note: this warning originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)
17
-       error[E0080]: erroneous constant used
+       error: erroneous constant encountered
19        --> $DIR/panic-assoc-never-type.rs:17:13
20         |
21      LL |     let _ = PrintName::VOID;

-          |             ^^^^^^^^^^^^^^^ referenced constant has errors
+          |             ^^^^^^^^^^^^^^^
23
24      error: aborting due to previous error; 1 warning emitted
25

-       For more information about this error, try `rustc --explain E0080`.
27
```

I'm unsure if leaving uninhabited ZSTs alone is the right thing to do or a hack that just happens to fix this case.

File: src/test/mir-opt/matches_reduce_branches.foo.PreCodegen.before.64bit.mir [hunk 2]
```diff
@@ -3,8 +3,10 @@
 fn foo(_1: Option<()>) -> () {
     debug bar => _1;                     // in scope 0 at $DIR/matches_reduce_branches.rs:7:8: 7:11
     let mut _0: ();                      // return place in scope 0 at $DIR/matches_reduce_branches.rs:7:25: 7:25
+    let mut _2: isize;                   // in scope 0 at $DIR/matches_reduce_branches.rs:8:22: 8:26
 
     bb0: {
-        return;                          // scope 0 at $DIR/matches_reduce_branches.rs:11:2: 11:2
+        _2 = discriminant(_1);           // scope 0 at $DIR/matches_reduce_branches.rs:8:22: 8:26
```
Comment (erikdesjardins) [inline src/test/mir-opt/matches_reduce_branches.foo.PreCodegen.before.64bit.mir:9]: It's weird that this comes back, we should be removing the ZST assignment even earlier than before.

File: compiler/rustc_mir/src/transform/instcombine.rs (hunk 1 above)
Comment (oli-obk) [reply compiler/rustc_mir/src/transform/instcombine.rs:53]: This change is fine

File: compiler/rustc_mir/src/transform/instcombine.rs [hunk 3]
```diff
@@ -41,6 +44,28 @@ impl<'tcx, 'a> InstCombineContext<'tcx, 'a> {
         })
     }
 
+    /// Remove assignments to inhabited ZST places.
+    fn combine_zst(&self, source_info: &SourceInfo, kind: &mut StatementKind<'tcx>) {
+        match kind {
+            StatementKind::Assign(box (place, _)) => {
+                let place_ty = place.ty(self.local_decls, self.tcx).ty;
+                if let Ok(layout) = self.tcx.layout_of(self.param_env.and(place_ty)) {
```
Comment (oli-obk) [inline compiler/rustc_mir/src/transform/instcombine.rs:52]: This `layout_of` invocation actually shows up in some of the regressions

File: compiler/rustc_mir/src/transform/instcombine.rs (hunk 1 above)
Comment (erikdesjardins) [reply compiler/rustc_mir/src/transform/instcombine.rs]: You mean the code is fine, or the error message change would be fine?

**Comment by erikdesjardins** [comment]
Pushed a change to cache layouts, let's see if it gets better or worse.

Also moved it to a separate pass, since it's a bit different than the other opts in instcombine...let me know if you have a preference for where it should live.

**Comment by tmiasko** [comment]
Looking at MIR diffs of some real world projects, this implementation is definitely more effective at removing ZST assignments than previous one was.  Though, it not demonstrated by any of existing mir-opt tests, so if we want to land this adding and extra one would be nice.

The perf results, both those here and earlies ones, are quite hard to interpret.  Unfortunately the most significant impact of this change is one on the size estimates. In a few benchmarks I looked at, the CGU partitioning was changed. This almost surely applies to rustc itself as well. In fact, I suspect that the -3.0% change in ctfe-stress-4 benchmark from earlier perf run were entirely because of this (code that is hot in those benchmarks is optimized differently and CTFE evaluates unoptimized MIR).

The layout computation uses query system, the computation should be cached already, but we can of course try again:

@bors try @rust-timer queue

File: compiler/rustc_mir/src/transform/instcombine.rs (hunk 1 above)
Comment (oli-obk) [reply compiler/rustc_mir/src/transform/instcombine.rs]: The change in the error message is totally alright

**Comment by oli-obk** [comment]
Deeply nested async still has lots of new query calls (layout_of and something around normalization). The CTFE stress tests regressed a bit, but that is fine, likely they have more instructions to execute? Though I wonder how this optimization affects CTFE as CTFE runs without optimizations.

**Comment by tmiasko** [comment]
<details><summary>We don't run SimplifyLocals in MIR used for CTFE</summary>

```rust
pub const fn f() {
    1;
}
```

```console
$ rustc --crate-type=lib z.rs -Zunpretty=mir
// WARNING: This output format is intended for human consumers only
// and is subject to change without notice. Knock yourself out.
fn f() -> () {
    let mut _0: ();                      // return place in scope 0 at z.rs:1:18: 1:18

    bb0: {
        return;                          // scope 0 at z.rs:3:2: 3:2
    }
}

// MIR FOR CTFE
fn f() -> () {
    let mut _0: ();                      // return place in scope 0 at z.rs:1:18: 1:18
    let _1: i32;                         // in scope 0 at z.rs:2:5: 2:6

    bb0: {
        StorageLive(_1);                 // scope 0 at z.rs:2:5: 2:6
        _1 = const 1_i32;                // scope 0 at z.rs:2:5: 2:6
        StorageDead(_1);                 // scope 0 at z.rs:2:6: 2:7
        _0 = const ();                   // scope 0 at z.rs:1:18: 3:2
        return;                          // scope 0 at z.rs:3:2: 3:2
    }
}
```
</details>

**Comment by oli-obk** [comment]
I meant the `simplify_locals` function, which is also called elsewhere, but checking that... it's also only run with other optimizations. So... this is another effect of rustc getting slower for CTFE when MIR opts change and not CTFE having less MIR to process...

File: compiler/rustc_mir/src/transform/remove_zsts.rs [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+impl<'tcx> MirPass<'tcx> for RemoveZsts {
+    fn run_pass(&self, tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
+        let param_env = tcx.param_env(body.source.def_id());
+        let (basic_blocks, local_decls) = body.basic_blocks_and_local_decls_mut();
+        for block in basic_blocks.iter_mut() {
+            for statement in block.statements.iter_mut() {
+                match statement.kind {
+                    StatementKind::Assign(box (place, _)) => {
+                        let place_ty = place.ty(local_decls, tcx).ty;
+                        if let Ok(layout) = tcx.layout_of(param_env.and(place_ty)) {
```
Comment (oli-obk) [inline compiler/rustc_mir/src/transform/remove_zsts.rs:18]: Maybe a fast path for known ZSTs (well, let's start with just `()`) could reduce the number of query calls?

**Comment by erikdesjardins** [comment]
Added a check to skip `layout_of` for types which can never be ZSTs.

When compiling std (or whatever gets built during a stage 1 build), the RemoveZsts pass now sees:
```
855924 total assignments

478400 assignments are skipped by the `maybe_zst` check
353160 assignments are skipped by the `layout_of` check
 24364 assignments are removed due to being of a ZST
```
I didn't add a fast path for known ZSTs because they make up <10% of the remaining `layout_of` calls. I can try that, or make the `maybe_zst` check more precise, if this isn't enough.

**Comment by oli-obk** [comment]
> ```
> 
> ```
> 
> 
> I didn't add a fast path for known ZSTs because they make up <10% of the remaining `layout_of` calls. I can try that, or make the `maybe_zst` check more precise, if this isn't enough.

so... 90% of zsts are aggregates or user defined? I would have thought a large portion is `()` or FnDefs

@bors try @rust-timer queue

**Comment by erikdesjardins** [comment]
What I assume you meant by "fast path" is
```rust
if ty == unit {
    // fast path
} else {
    // slow path
    if let Ok(layout) = layout_of(ty) && layout.is_zst() {
        // slow path success
    } else {
        // slow path failure
    }
}
```
In my test, the fast path could be hit at most 24k times, if every ZST is `()`. But the slow path would still be hit at least 353k times, because there are 353k assignments that aren't ZSTs, but aren't ruled out until we check the layout (i.e., we reach "slow path failure" at least 353k times). I don't expect a fast path that's hit <10% of the time (24k / [24k + 353k] ~ 6%) to significantly improve performance.

Unless you meant "add a fast path and remove the slow path entirely", i.e. the optimization only works for `()`, FnDef, etc., and not `struct MyZst;`, but I'd prefer not to do that if possible.

**Comment by oli-obk** [comment]
> Unless you meant "add a fast path and remove the slow path entirely", i.e. the optimization only works for `()`, FnDef, etc., and not `struct MyZst;`, but I'd prefer not to do that if possible.

I did not mean that. My brain just took a wrong turn somewhere. You're completely right.

Though... we could enable the optimization for FnDef and unit in debug builds and for everything in release builds, but let's look at perf before we resort to such schemes.

**Comment by oli-obk** [comment]
Perf looks very promising. While there's still some regression in servo, that is entirely in LLVM, so we may be optimizing more stuff now, no way to tell without runtime perf tests. Also the LLVM perf test shows a 60% reduction in `static_mutability` query calls (15k!!!) on the servo test.

@erikdesjardins this looks really good, all that is left is to add a mir-opt-level 3 check in the opt so it doesn't run by default. I think we should do that here and not immediately stabilize, even if I see no reason not to stabilize. The opt doesn't affect anything UB related and is trivial to review. So my proposal is to merge this PR quickly with a level 3 check, and then open a PR removing that check and pinging wg-mir-opt so that everyone can have their say

**Pull request merged**